File: Computer networks/socket/dns_lookup/DNSLookup.cpp
#include <netinet/in.h> 
#include <sys/socket.h>
#include <sys/types.h>
#include <arpa/inet.h>
#include <iostream>
#include <netdb.h>
#include <cstring>
#include <string>

struct in_addr IPv4addr;
struct hostent *host;

int main(int argc, char** argv) {
	char hostname[50] = "localhost";
	bool localhost = false;

	if(argc == 1) {
		host = gethostbyname(hostname);
		localhost = true;
	}

	if (argc > 2) {
		std::cout << std::endl << "Usage :" << std::endl;
		std::cout << "./DNSLookup [hostname/IP Address]" << std::endl << std::endl;
		return(1);
	}

	if(!localhost) {
		std::strcpy(hostname, argv[1]);
		if(hostname[0] >= 48 && hostname[0] <= 57) {
			inet_pton(AF_INET, hostname, &IPv4addr);
			host = gethostbyaddr(&IPv4addr, sizeof(IPv4addr), AF_INET);   //get hostent by IP address
		}
		else
			host = gethostbyname(hostname);  //get hostent by  domain name

		if(host == NULL) { 
			std::cout << "undefined hostname : " <<  hostname << std::endl; 
			return(2);
		}
	}

	std::cout << "------- Hostname info -------" << std::endl << std::endl;
	std::cout << "Name : " <<  host->h_name << std::endl;
	
	for (int i = 0; host->h_aliases[i]; ++i)
		std::cout << "Aliases : " << host->h_aliases[i] << std::endl;

	std::cout << "Address type : " << host->h_addrtype << std::endl;
	std::cout << "Length : " << host->h_length << std::endl;
	std::cout << "Address : " << inet_ntoa(*(struct in_addr*)host->h_addr) << std::endl;

	return (0);
}

File: Computer networks/socket/dns_lookup/DNSLookup.py
#!/usr/bin/python

import socket
import sys

def isInt(str):
	try:
		int(str)
		return True
	except ValueError:
		return False


def dnsLookup(hostname):
	print "------- Hostname info -------"
	print ""

	if isInt(sys.argv[1][0]) == True:
		print "Name : " + socket.gethostbyaddr(hostname)[0]
		for als in socket.gethostbyaddr(hostname)[1]:
			print "Alias : " + als

		print "Address type : " + str(socket.getaddrinfo(hostname, 80)[0][1])
		for ips in socket.gethostbyaddr(hostname)[2]:
			print "Address : " + ips

	else:
		print "Name : " + socket.gethostbyname_ex(hostname)[0]
		for als in socket.gethostbyname_ex(hostname)[1]:
			print "Alias : " + als

		print "Address type : " + str(socket.getaddrinfo(hostname, 80)[0][1])
		for ips in socket.gethostbyname_ex(hostname)[2]:
			print "Address : " + ips

if len(sys.argv) == 1:
	dnsLookup("localhost")

if len(sys.argv) == 2:
	try:
		dnsLookup(sys.argv[1])
	except:
		print "unknow hostname"
		SystemExit("unknow argument")

if len(sys.argv) > 2:
	print "Usage : "
	print "python DNSLookup [hostname/IP]"
	SystemExit("1 argument requested, 2 given")

